File: Swift/47 Permutations II.swift
class Solution {
    func permuteUnique(_ nums: [Int]) -> [[Int]] {
        var ans: [[Int]] = []
        var path: [Int] = []
        let res = nums.sorted()

        let n = res.count
        var state: [Bool] = .init(repeating: false, count: n)
        func dfs(_ cnt: Int) {
            if cnt == res.count {
                ans.append(path)
                return
            }

            for i in 0 ..< res.count {
                if state[i] {
                    continue
                }

                if i > 0, res[i - 1] == res[i], state[i - 1] {
                    continue
                }

                state[i] = true
                path.append(res[i])
                dfs(cnt + 1)
                path.popLast()
                state[i] = false
            }
        }
        dfs(0)
        return ans
    }
}

File: C++/47 Permutations II.cpp
class Solution
{
public:
    vector<vector<int>> permuteUnique(vector<int> &nums)
    {
        vector<vector<int>> ans;
        vector<int> path;
        sort(nums.begin(), nums.end());

        int n = nums.size();
        vector<bool> state(n, false);
        auto dfs = [&](this auto &&dfs, int cnt) -> void
        {
            if (cnt == n)
            {
                ans.push_back(path);
                return;
            }

            for (int i = 0; i < n; i++)
            {
                if (state[i])
                    continue;
                if (i > 0 && nums[i] == nums[i - 1] && state[i - 1])
                    continue;
                state[i] = true;
                path.push_back(nums[i]);
                dfs(cnt + 1);
                path.pop_back();
                state[i] = false;
            }
        };
        dfs(0);
        return ans;
    }
};